lib/Dialect/Arc/Transforms/LowerState.cpp

```diff
@@ -0,0 +1,605 @@
+        blockArg.getLoc(), StateType::get(intType), name, storageArg);
+    replaceValueWithStateRead(blockArg, state);
+  }
+  return success();
+}
+
+/// Lower the primary outputs of the module to dedicated ops that allocate the
+/// outputs in the model's storage.
+LogicalResult ModuleLowering::lowerPrimaryOutputs() {
+  auto outputOp = cast<hw::OutputOp>(moduleOp.getBodyBlock()->getTerminator());
+  if (outputOp.getNumOperands() > 0) {
+    auto &passThrough = getOrCreatePassThrough();
+    for (auto [value, name] :
+         llvm::zip(outputOp.getOperands(), moduleOp.getResultNames())) {
+      auto intType = value.getType().dyn_cast<IntegerType>();
+      if (!intType)
+        return mlir::emitError(outputOp.getLoc(), "output ")
+               << name << " is of non-integer type " << value.getType();
+      auto materializedValue = passThrough.materializeValue(value);
+      auto state = builder.create<RootOutputOp>(
+          outputOp.getLoc(), StateType::get(intType), name.cast<StringAttr>(),
+          storageArg);
+      passThrough.builder.create<StateWriteOp>(outputOp.getLoc(), state,
+                                               materializedValue, Value{});
+    }
+  }
+  return success();
+}
+
+LogicalResult ModuleLowering::lowerStates() {
+  for (auto &op : llvm::make_early_inc_range(*moduleOp.getBodyBlock())) {
+    auto result = TypeSwitch<Operation *, LogicalResult>(&op)
+                      .Case<StateOp, MemoryOp, MemoryWriteOp, TapOp>(
+                          [&](auto op) { return lowerState(op); })
+                      .Default(success());
+    if (failed(result))
+      return failure();
+  }
+  return success();
+}
+
+LogicalResult ModuleLowering::lowerState(StateOp stateOp) {
+  // Latency zero arcs incur no state and remain in the IR unmodified.
+  if (stateOp.getLatency() == 0)
+    return success();
+
+  // We don't support arcs beyond latency 1 yet. These should be easy to add in
+  // the future though.
+  if (stateOp.getLatency() > 1) {
+    stateOp.emitError("state with latency > 1 not supported");
+    return failure();
+  }
+
+  // Arcs with latency must be synchronous to a clock.
+  if (!stateOp.getClock()) {
+    stateOp.emitError("state with latency > 0 requires a clock");
+    return failure();
+  }
+
+  // Get the clock tree and enable condition for this state's clock. If this arc
+  // carries an explicit enable condition, fold that into the enable provided by
+  // the clock gates in the arc's clock tree.
+  auto info = getOrCreateClockLowering(stateOp.getClock());
+  info.enable = info.clock.getOrCreateAnd(info.enable, stateOp.getEnable(),
+                                          stateOp.getLoc());
+
+  // Allocate the necessary state within the model.
+  SmallVector<Value> allocatedStates;
+  for (unsigned stateIdx = 0; stateIdx < stateOp.getNumResults(); ++stateIdx) {
+    auto stateType = StateType::get(
+        stateOp.getResult(stateIdx).getType().cast<IntegerType>());
```

Comment:
The `arc.state` ops results are not verified to be integers, so it might also be an array and crash here.

Reply:
Good point. I've changed this to emit an error for now, but we'll probably want to change this to be able to support other types as well. Or have a pre-pass that lowers arc uses to trivial integers.